common/primitive/atomic_u128.rs

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (C) 2023 Vaticle
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+use std::sync::atomic::AtomicU64;
+use std::sync::atomic::Ordering::SeqCst;
+
+/// Implements an AtomicU128 restricted to increments of u32.
+/// The lower-word (LW) is implemented using an AtomicU64. It is always correct.
+/// The upper word (UW) is implemented using u128, with an overlap of the upper 8 bits of the lower-word. The upper word + 8 sync-bits are eventually correct.
+/// The overlap is used to detect the UW being outdated.
+/// We restrict increments to at most u32::MAX, so that it is reasonable to assume:
+/// 1. there are never concurrent updates to the sync-bits or to UW.
+/// 2. The UW is out of sync by exactly 1 when sync_bits(LW) < sync_bits(UW), and is up-to-date otherwise
+///
+/// It is the responsibility of the thread that updates the sync_bits (and possibly overflow) to update sync_bits in the u128.
+/// Other threads can return (UW+1)(LW) if sync_bits(LW) < sync_bits(UW);  (UW)(LW) otherwise
+pub struct AtomicU128 {
+    uw_sync: u128,
+    lw: AtomicU64,
+}
+
+const SYNC_MASK: u64 = 0xff00000000000000;
+const UW_MASK: u128 = 0xffffffffffffffff_0000000000000000;
+const UW_INCREMENT: u128 = 0x0000000000000001_0000000000000000;
```

Comment:
I'd suggest breaking up 64- and 128-bit literals further, so it's obvious the length is correct.
```
0xff00_0000_0000_0000
0xffff_ffff_ffff_ffff__0000_0000_0000_0000
```